File: app/src/main/java/com/example/myapplication/codingminutes/divide_conquer/BinarySearch.java
package com.example.myapplication.codingminutes.divide_conquer;

public class BinarySearch {

    public static void main(String[] args) {
        int[] arr = {1, 3, 5, 7, 9};
        int x = 9;
        int s = 0;
        int e = arr.length - 1;
        int result = binarySearch(arr, s, e, x);
        System.out.println("INDEX::" + result);
    }

    static int binarySearch(int[] arr, int s, int e, int x) {
        //base case
        if (e >= s) {
            int mid = s + (e - s) / 2;
            if (arr[mid] == x) {
                return mid;
            }
            if (x > arr[mid]) {
                return binarySearch(arr, mid + 1, e, x);
            } else {
                return binarySearch(arr, s, mid - 1, x);
            }
        }
        return -1;
    }
}

File: app/src/main/java/com/example/myapplication/codingminutes/array/BinarySearch.kt
package com.example.myapplication.codingminutes.array

fun main() {
    val array = arrayOf(1, 6, 14, 16, 22, 24, 26, 33, 35, 44)
    val result = binarySearch(array, 3)
    println("RESULT::$result")
}

fun binarySearch(array: Array<Int>, key: Int): Int {
    var start = 0
    var end = array.size - 1

    while (start <=end) {
        var mid = (start + end) / 2
        if (array[mid] == key) {
            return mid
        } else if (key < array[mid]) {
            end = mid - 1
        } else {
            start = mid + 1
        }
    }
    return -1
}